# **Sub Division population of pakistan** 

- This data is not mine I just download it from kaggle.
- I am practing & improving my skills 
- So that's it 

- Let's begin

## Import libraries

In [2]:
import pandas as pd
import seaborn as sns
import matplotlib.pyplot as plt
#this is for jupyter notebook to show the plot in the notebook itself instead of opening a new window for the plot.
%matplotlib inline 
import numpy as np
import plotly.express as px

- Let's Loading the data and creating df.

In [3]:
df = pd.read_csv('sub-division_population_of_pakistan.csv')

- Let's have look on data

In [4]:
df.head(5)

PROVINCE             DIVISION               DISTRICT         SUB DIVISION  \
0   PUNJAB  BAHAWALPUR DIVISION  BAHAWALNAGAR DISTRICT  BAHAWALNAGAR TEHSIL   
1   PUNJAB  BAHAWALPUR DIVISION  BAHAWALNAGAR DISTRICT     CHISHTIAN TEHSIL   
2   PUNJAB  BAHAWALPUR DIVISION  BAHAWALNAGAR DISTRICT    FORT ABBAS TEHSIL   
3   PUNJAB  BAHAWALPUR DIVISION  BAHAWALNAGAR DISTRICT    HAROONABAD TEHSIL   
4   PUNJAB  BAHAWALPUR DIVISION  BAHAWALNAGAR DISTRICT   MINCHINABAD TEHSIL   

   AREA (sq.km)  ALL SEXES (RURAL)  MALE (RURAL)  FEMALE (RURAL)  \
0        1729.0             619550        316864          302644   
1        1500.0             540342        273788          266500   
2        2536.0             361240        182655          178541   
3        1295.0             382115        192278          189808   
4        1818.0             452723        231506          221178   

   TRANSGENDER (RURAL)  SEX RATIO (RURAL)  ...  POPULATION 1998 (RURAL)  \
0                   42             104.70  ...                   407768   
1                   54             102.73  ...                   395983   
2                   44             102.30  ...                   250959   
3                   29             101.30  ...                   297343   
4                   39             104.67  ...                   316593   

   ANNUAL GROWTH RATE (RURAL)  ALL SEXES (URBAN)  MALE (URBAN)  \
0                        2.22             193840         98391   
1                        1.65             149424         75546   
2                        1.93              61528         31360   
3                        1.33             142600         71345   
4                        1.90              72294         36694   

   FEMALE (URBAN)  TRANSGENDER (URBAN)  SEX RATIO (URBAN)  \
0           95402                   47             103.13   
1           73851                   27             102.30   
2           30150                   18             104.01   
3           71236                   19             100.15   
4           35590                   10             103.10   

   AVG HOUSEHOLD SIZE (URBAN)  POPULATION 1998 (URBAN)  \
0                        6.02                   133785   
1                        6.01                   102287   
2                        6.00                    34637   
3                        6.02                    84424   
4                        6.34                    37668   

   ANNUAL GROWTH RATE (URBAN)  
0                        1.97  
1                        2.01  
2                        3.06  
3                        2.79  
4                        3.48  

[5 rows x 21 columns]

> **Note**: Some the output of notebook does not present the complete output, therefore we can increase the limit of columns view and row view by using these commands: 

In [5]:
pd.set_option('display.max_columns', None) # this is to display all the columns in the dataframe
pd.set_option('display.max_rows', None) # this is to display all the rows in the dataframe

In [6]:
# hide all warnings runtime
import warnings
warnings.filterwarnings('ignore')

- Let's see the column names.

In [7]:
df.columns

Index(['PROVINCE', 'DIVISION', 'DISTRICT', 'SUB DIVISION', 'AREA (sq.km)',
       'ALL SEXES (RURAL)', 'MALE (RURAL)', 'FEMALE (RURAL)',
       'TRANSGENDER (RURAL)', 'SEX RATIO (RURAL)',
       'AVG HOUSEHOLD SIZE (RURAL)', 'POPULATION 1998 (RURAL)',
       'ANNUAL GROWTH RATE (RURAL)', 'ALL SEXES (URBAN)', 'MALE (URBAN)',
       'FEMALE (URBAN)', 'TRANSGENDER (URBAN)', 'SEX RATIO (URBAN)',
       'AVG HOUSEHOLD SIZE (URBAN)', 'POPULATION 1998 (URBAN)',
       'ANNUAL GROWTH RATE (URBAN)'],
      dtype='object')

- Let's check the exact rows & columns of dataset.

In [8]:
print(f'The number of rows in the dataset is {df.shape[0]} & the number of columns in the dataset is {df.shape[1]}.')

The number of rows in the dataset is 528 & the number of columns in the dataset is 21.


- I can't understand the data Let's check the datatypes and info by using info function

In [9]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 528 entries, 0 to 527
Data columns (total 21 columns):
 #   Column                      Non-Null Count  Dtype  
---  ------                      --------------  -----  
 0   PROVINCE                    528 non-null    object 
 1   DIVISION                    528 non-null    object 
 2   DISTRICT                    528 non-null    object 
 3   SUB DIVISION                528 non-null    object 
 4   AREA (sq.km)                528 non-null    float64
 5   ALL SEXES (RURAL)           528 non-null    int64  
 6   MALE (RURAL)                528 non-null    int64  
 7   FEMALE (RURAL)              528 non-null    int64  
 8   TRANSGENDER (RURAL)         528 non-null    int64  
 9   SEX RATIO (RURAL)           528 non-null    float64
 10  AVG HOUSEHOLD SIZE (RURAL)  528 non-null    float64
 11  POPULATION 1998 (RURAL)     528 non-null    int64  
 12  ANNUAL GROWTH RATE (RURAL)  528 non-null    float64
 13  ALL SEXES (URBAN)           528 non

# **Observations**
---
1. There are 528 rows and 21 columns in the dataset
2. The columns in the datasets are:
 - These are object columns:
    - `'PROVINCE', 'DIVISION', 'DISTRICT', 'SUB DIVISION'`
 - These are numeric columns:
    - ` 'AREA (sq.km)',
       'ALL SEXES (RURAL)', 'MALE (RURAL)', 'FEMALE (RURAL)',
       'TRANSGENDER (RURAL)', 'SEX RATIO (RURAL)',
       'AVG HOUSEHOLD SIZE (RURAL)', 'POPULATION 1998 (RURAL)',
       'ANNUAL GROWTH RATE (RURAL)', 'ALL SEXES (URBAN)', 'MALE (URBAN)',
       'FEMALE (URBAN)', 'TRANSGENDER (URBAN)', 'SEX RATIO (URBAN)',
       'AVG HOUSEHOLD SIZE (URBAN)', 'POPULATION 1998 (URBAN)',
       'ANNUAL GROWTH RATE (URBAN)'`
  

In [10]:
df.describe().T

count           mean            std  min  \
AREA (sq.km)                528.0    1492.005871    2039.453778  0.0   
ALL SEXES (RURAL)           528.0  246278.011364  271189.816559  0.0   
MALE (RURAL)                528.0  125275.687500  137563.021458  0.0   
FEMALE (RURAL)              528.0  120984.149621  133716.898296  0.0   
TRANSGENDER (RURAL)         528.0      18.174242      25.522248  0.0   
SEX RATIO (RURAL)           528.0      98.982614      26.812660  0.0   
AVG HOUSEHOLD SIZE (RURAL)  528.0       6.277064       2.074947  0.0   
POPULATION 1998 (RURAL)     528.0  167427.994318  178388.976993  0.0   
ANNUAL GROWTH RATE (RURAL)  528.0       3.124792       9.577872  0.0   
ALL SEXES (URBAN)           528.0  140863.528409  351246.279136  0.0   
MALE (URBAN)                528.0   72843.386364  182349.241379  0.0   
FEMALE (URBAN)              528.0   67997.865530  168872.528212  0.0   
TRANSGENDER (URBAN)         528.0      22.276515      66.068127  0.0   
SEX RATIO (URBAN)           528.0      75.411269      49.687341  0.0   
AVG HOUSEHOLD SIZE (URBAN)  528.0       4.406402       2.948336  0.0   
POPULATION 1998 (URBAN)     528.0   80144.566288  202312.019227  0.0   
ANNUAL GROWTH RATE (URBAN)  528.0       1.920814       2.098908  0.0   

                                  25%         50%          75%         max  
AREA (sq.km)                  425.000     882.000    1734.2500    18374.00  
ALL SEXES (RURAL)           50934.500  165241.000  312911.2500  2297375.00  
MALE (RURAL)                27127.250   84134.500  160502.2500  1172995.00  
FEMALE (RURAL)              23979.000   82044.000  152219.0000  1124167.00  
TRANSGENDER (RURAL)             1.000      10.000      25.0000      213.00  
SEX RATIO (RURAL)             101.105     105.285     108.3475      139.38  
AVG HOUSEHOLD SIZE (RURAL)      5.700       6.310       7.2000       12.43  
POPULATION 1998 (RURAL)     35273.500  117206.500  213054.2500  1044035.00  
ANNUAL GROWTH RATE (RURAL)      1.440       2.030       2.8125      100.00  
ALL SEXES (URBAN)               0.000   43254.500  117814.7500  3653616.00  
MALE (URBAN)                    0.000   21980.000   60301.7500  1905921.00  
FEMALE (URBAN)                  0.000   20999.000   57465.2500  1746900.00  
TRANSGENDER (URBAN)             0.000       3.000      19.0000      795.00  
SEX RATIO (URBAN)               0.000     103.195     107.0400      297.81  
AVG HOUSEHOLD SIZE (URBAN)      0.000       5.755       6.3400       10.06  
POPULATION 1998 (URBAN)         0.000   21298.000   65422.2500  2075867.00  
ANNUAL GROWTH RATE (URBAN)      0.000       1.855       2.9850       19.78

## **2.2. Dealing with the missing values**
Dealing with the missing values is one of the most important part of the data wrangling process, we must deal with the missing values in order to get the correct insights from the data.

- Let's check the missing values in the dataset

In [11]:
df.isnull().sum()

PROVINCE                      0
DIVISION                      0
DISTRICT                      0
SUB DIVISION                  0
AREA (sq.km)                  0
ALL SEXES (RURAL)             0
MALE (RURAL)                  0
FEMALE (RURAL)                0
TRANSGENDER (RURAL)           0
SEX RATIO (RURAL)             0
AVG HOUSEHOLD SIZE (RURAL)    0
POPULATION 1998 (RURAL)       0
ANNUAL GROWTH RATE (RURAL)    0
ALL SEXES (URBAN)             0
MALE (URBAN)                  0
FEMALE (URBAN)                0
TRANSGENDER (URBAN)           0
SEX RATIO (URBAN)             0
AVG HOUSEHOLD SIZE (URBAN)    0
POPULATION 1998 (URBAN)       0
ANNUAL GROWTH RATE (URBAN)    0
dtype: int64

## **There are no null values in th dataset**
###  **That's great**

## **Lets's check the duplicates in the data**

In [12]:
df.duplicated().sum()

np.int64(0)

## There are no duplicates
- I think this data was clean 
  

# Let's Find the insights in the data.

- Let's see the province 

In [76]:
print(f'These are the provinces {df['PROVINCE'].unique()}')

These are the provinces ['PUNJAB' 'SINDH' 'BALOCHISTAN' 'KPK' 'KPK/FATA']


- Let's see the divisions

In [84]:
print(f'These are the divisions: \n => {df['DIVISION'].unique()}')

These are the divisions: 
 => ['BAHAWALPUR DIVISION' 'D.G.KHAN DIVISION' 'FAISALABAD DIVISION'
 'GUJRANWALA DIVISION' 'LAHORE DIVISION' 'MULTAN DIVISION'
 'RAWALPINDI DIVISION' 'SAHIWAL DIVISION' 'SARGODHA DIVISION'
 'Badin Division' 'Hyderabad Division' 'Karachi Division'
 'Larkana Division' 'Mirpurkhas Division' 'Shaheed Benazirabad Division'
 'Sukkur Division' 'Makran Division' 'Kalat Division'
 'Naseerabad Division' 'Quetta Division' 'Zhob Division' 'BANNU DIVISION'
 'DERA ISMAIL KHAN DIVISION' 'HAZARA DIVISION' 'KOHAT DIVISION'
 'MARDAN DIVISION' 'PESHAWAR DIVISION' 'MALAKAND DIVISION']


- Find Top 5 Sub-divison in rural-pop-1998

In [13]:
df.nlargest(5,'POPULATION 1998 (RURAL)')[['SUB DIVISION','POPULATION 1998 (RURAL)']]

SUB DIVISION  POPULATION 1998 (RURAL)
455          PESHAWAR TEHSIL                  1044035
35          JARANWALA TEHSIL                   924849
34   FAISALABAD SADAR TEHSIL                   900984
24       MUZAFFARGARH TEHSIL                   842563
125           SAHIWAL TEHSIL                   826863

- Find Top 5 Sub-divison in urban-pop-1998

In [14]:
df.nlargest(5,'POPULATION 1998 (URBAN)')[['SUB DIVISION','POPULATION 1998 (RURAL)']]

SUB DIVISION  POPULATION 1998 (RURAL)
71         LAHORE CITY TEHSIL                   143532
33     FAISALABAD CITY TEHSIL                   131485
117         RAWALPINDI TEHSIL                   517844
74            SHALIMAR TEHSIL                   201105
47   GUJRANWALA SADDAR TEHSIL                   586535

- Find Top 5 DISTRICT in urban-pop-1998

In [15]:
df.nlargest(5,'POPULATION 1998 (URBAN)')[['DISTRICT','POPULATION 1998 (RURAL)']]

DISTRICT  POPULATION 1998 (RURAL)
71       LAHORE DISTRICT                   143532
33   FAISALABAD DISTRICT                   131485
117  RAWALPINDI DISTRICT                   517844
74       LAHORE DISTRICT                   201105
47   GUJRANWALA DISTRICT                   586535

- Find Top 5 DISTRICT in rural-pop-1998

In [16]:
df.nlargest(5,'POPULATION 1998 (RURAL)')[['DISTRICT','POPULATION 1998 (RURAL)']]

DISTRICT  POPULATION 1998 (RURAL)
455      PESHAWAR DISTRICT                  1044035
35     FAISALABAD DISTRICT                   924849
34     FAISALABAD DISTRICT                   900984
24   MUZAFFARGARH DISTRICT                   842563
125       SAHIWAL DISTRICT                   826863

- Total population  in rural  & urban

In [43]:
print(f'Total population in rural:{df['POPULATION 1998 (RURAL)'].sum()} & in urban:{df['POPULATION 1998 (URBAN)'].sum()}')
print(f'And the overall total population is {(df['POPULATION 1998 (RURAL)'].sum()) + (df['POPULATION 1998 (URBAN)'].sum())}')

Total population in rural:88401981 & in urban:42316331
And the overall total population is 130718312


- The average population of cities in each province.

In [57]:
print('The average population in cities each province.')
print(f"=> RURAL : {df.groupby('PROVINCE')['POPULATION 1998 (RURAL)'].mean()}")
print(f"=> URBAN : {df.groupby('PROVINCE')['POPULATION 1998 (URBAN)'].mean()}")
print(f"=> IN BOTH : {(df.groupby('PROVINCE')['POPULATION 1998 (RURAL)'].mean()) + (df.groupby('PROVINCE')['POPULATION 1998 (URBAN)'].mean())}")

The average population in cities each province.
=> RURAL : PROVINCE
BALOCHISTAN     37944.450382
KPK            209638.816901
KPK/FATA        62657.086957
PUNJAB         350172.034965
SINDH          113796.167883
Name: POPULATION 1998 (RURAL), dtype: float64
=> URBAN : PROVINCE
BALOCHISTAN     11975.419847
KPK             40732.830986
KPK/FATA         1858.108696
PUNJAB         160350.944056
SINDH          108320.160584
Name: POPULATION 1998 (URBAN), dtype: float64
=> IN BOTH : PROVINCE
BALOCHISTAN     49919.870229
KPK            250371.647887
KPK/FATA        64515.195652
PUNJAB         510522.979021
SINDH          222116.328467
dtype: float64


 - Total divisions in each province.


In [61]:
df['PROVINCE'].value_counts()


PROVINCE
PUNJAB         143
SINDH          137
BALOCHISTAN    131
KPK             71
KPK/FATA        46
Name: count, dtype: int64

- Total districts in each division

In [62]:
df['DIVISION'].value_counts()

DIVISION
Quetta Division                 38
Makran Division                 37
MALAKAND DIVISION               33
Karachi Division                30
Hyderabad Division              30
RAWALPINDI DIVISION             22
Larkana Division                21
GUJRANWALA DIVISION             20
Naseerabad Division             19
Zhob Division                   19
Sukkur Division                 18
Mirpurkhas Division             18
PESHAWAR DIVISION               18
Kalat Division                  18
LAHORE DIVISION                 17
SARGODHA DIVISION               17
FAISALABAD DIVISION             17
BAHAWALPUR DIVISION             15
Shaheed Benazirabad Division    15
BANNU DIVISION                  15
HAZARA DIVISION                 15
DERA ISMAIL KHAN DIVISION       15
MULTAN DIVISION                 14
D.G.KHAN DIVISION               14
KOHAT DIVISION                  14
SAHIWAL DIVISION                 7
MARDAN DIVISION                  7
Badin Division                   5
Name: count

- Average annual growth rate in rural , urban and overall.

In [95]:
print(f'Average annual growth rate in rural {df['ANNUAL GROWTH RATE (RURAL)'].mean()}%.')
print(f'Average annual growth rate in urban {df['ANNUAL GROWTH RATE (URBAN)'].mean()}%.')
print(f'Average annual growth rate {(df['ANNUAL GROWTH RATE (RURAL)'].mean())+(df['ANNUAL GROWTH RATE (URBAN)'].mean())}%.')


Average annual growth rate in rural 3.1247916666666664%.
Average annual growth rate in urban 1.920814393939394%.
Average annual growth rate 5.04560606060606%.


- Average sex ratio in rural , urban and overall.

In [110]:
print(f"Average sex ratio in rural {df["SEX RATIO (RURAL)"].mean()}")
print(f"Average sex ratio in urban {df["SEX RATIO (URBAN)"].mean()}")
print(f"Average sex ratio overall {(df["SEX RATIO (URBAN)"].mean())+(df["SEX RATIO (RURAL)"].mean())}")

Average sex ratio in rural 98.98261363636364
Average sex ratio in urban 75.41126893939395
Average sex ratio overall 174.39388257575757


- Average Houswhold size in rural , urban and overalll

In [118]:
print(f'Average household size in rural {df['AVG HOUSEHOLD SIZE (RURAL)'].mean()}')
print(f'Average household size in urban {df['AVG HOUSEHOLD SIZE (URBAN)'].mean()}')
print(f'Average household size overall {(df['AVG HOUSEHOLD SIZE (URBAN)'].mean())+(df['AVG HOUSEHOLD SIZE (RURAL)'].mean())}')

Average household size in rural 6.277064393939394
Average household size in urban 4.406401515151515
Average household size overall 10.68346590909091
